pyccl/concentration.py

```diff
@@ -0,0 +1,317 @@
+
+    def _default_mdef(self):
+        self.mdef = MassDef(200, 'critical')
+
+    def _check_mdef(self, mdef):
+        if (np.fabs(mdef.Delta - 200.) > 1E-4) and \
+           (mdef.rho_type != 'critical'):
+            return True
+        return False
+
+    def _setup(self):
+        self.c0 = 3.681
+        self.c1 = 5.033
+        self.al = 6.948
+        self.x0 = 0.424
+        self.i0 = 1.047
+        self.i1 = 1.646
+        self.be = 7.386
+        self.x1 = 0.526
+        self.cnorm = 1. / self.cmin(1.393)
+        self.inorm = 1. / self.imin(1.393)
+
+    def cmin(self, x):
+        return self.c0 + (self.c1 - self.c0) * \
+            (np.arctan(self.al * (x - self.x0)) / np.pi + 0.5)
+
+    def imin(self, x):
+        return self.i0 + (self.i1 - self.i0) * \
+            (np.arctan(self.be * (x - self.x1)) / np.pi + 0.5)
+
+    def concentration(self, cosmo, M, a):
+        sig = sigmaM(cosmo, M, a)
+        om = cosmo.cosmo.params.Omega_m
+        ol = cosmo.cosmo.params.Omega_l
+        x = a * (ol / om)**(1. / 3.)
+        B0 = self.cmin(x) * self.cnorm
+        B1 = self.imin(x) * self.inorm
+        sig_p = B1 * sig
+        Cc = 2.881 * ((sig_p / 1.257)**1.022 + 1) * np.exp(0.060 / sig_p**2)
+        return B0 * Cc
+
+
+class ConcentrationKlypin11(Concentration):
+    name = 'Klypin11'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationKlypin11, self).__init__("Klypin11",
+                                                    mdef)
+
+    def _default_mdef(self):
+        self.mdef = MassDef('vir', 'critical')
+
+    def _check_mdef(self, mdef):
+        if mdef.Delta != 'vir':
+            return True
+        return False
+
+    def concentration(self, cosmo, M, a):
+        M_pivot_inv = cosmo.cosmo.params.h * 1E-12
+        return 9.6 * (M * M_pivot_inv)**-0.075
+
+
+class ConcentrationDuffy08(Concentration):
+    name = 'Duffy08'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationDuffy08, self).__init__("Duffy08",
+                                                   mdef)
+
+    def _default_mdef(self):
+        self.mdef = MassDef(200, 'critical')
+
+    def _check_mdef(self, mdef):
+        if mdef.Delta != 'vir':
+            if np.fabs(mdef.Delta - 200.) > 1E-4:
+                return True
+        return False
+
+    def _setup(self):
+        if self.mdef.Delta == 'vir':
+            self.A = 7.85
+            self.B = -0.081
+            self.C = -0.47
+        else:  # Now Delta has to be 200
+            if self.mdef.rho_type == 'matter':
+                self.A = 10.14
+                self.B = -0.081
+                self.C = -1.01
+            else:  # Now rho_type has to be critical
+                self.A = 5.71
+                self.B = -0.084
+                self.C = -0.47
+
+    def concentration(self, cosmo, M, a):
+        M_pivot_inv = cosmo.cosmo.params.h * 5E-13
+        return self.A * (M * M_pivot_inv)**self.B * a**(-self.C)
+
+
+def concentration_from_name(name):
```

Comment:
Do we need this factor function? Seems like the names are in the class names so that we are not saving anyone any effort?

Reply:
Meh, I thought this would be a convenient function to have. Saves O(10) characters each time, and it's also nice for the mass definitions (since you attach a c(M) to them just by passing the name of the c(M) parametrization).